### PR TITLE
reference/supply-chain: Add document on supply chain security / SLSA levels

### PR DESCRIPTION
**Pull request by t-lo**
This change adds a document on Flatcar's supply chain security and match respective SLSA levels.

**NOTE** that statements in this doc are only made with respect to the new "ci-automation" pipeline and containerised builds. The legacy pipeline has a significantly lower level of SLSA compliance.

See https://github.com/flatcar-linux/flatcar-docs/blob/t-lo/supply-chain-docs/docs/reference/supply-chain.md for a preview.

Fixes https://github.com/flatcar-linux/Flatcar/issues/771 .

### CODE REVIEW

File: docs/reference/supply-chain.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+
+### Flatcar Container Linux Supply Chain Security and SLSA
+
+The [Supply Chain Levels for Software Artifacts](https://slsa.dev/) (SLSA or 'salsa' in short) defines a check-list of standards and controls to prevent tampering, improve integrity, and secure packages and infrastructure in software projects.
+We lay out below how the Flatcar Container Linux project works towards implementing the [requirements of SLSA](https://slsa.dev/spec/v0.1/requirements) and provide a deep dive into the processes and mechanisms to secure our supply chain in the Flatcar project.
+
+**Note** while Flatcar implements most of the requirements defined by SLSA we no not at this point generate user-verifiable [provenance](https://slsa.dev/provenance/v0.2) of doing so. This is planned as a future improvement to our release process.
+
+#### SLSA Thread model and requirements
```
Comment (marga-kinvolk) [inline docs/reference/supply-chain.md:14]: This should be Threat not Thread

File: docs/reference/supply-chain.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+
+### Deep dive: Implementation in Flatcar Container Linux
+
+Flatcar Container Linux employs a number of concrete mechanisms and processes to secure our supply chain.
+
+Broadly speaking, these break down into two areas:
+1. Mechanisms and processes to ensure validity of the Flatcar artifacts that make up our releases.
+   Attestation is performed either automatically by the build system or by the maintainers team.
+   This includes validating our inputs / upstreams, securing our builds, and ensuring attestability of our images and update payloads.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:62]:
```suggestion
   This includes validating the build pipeline's inputs / upstreams, securing the build process, and ensuring attestability of the resultant images and update payloads.
```

File: docs/reference/supply-chain.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+2. Mechanisms and processes to be applied at provisioning time by our users, as well as automatically at runtime, attesting validity of the artifacts in use.
+   Attestation is performed either automatically by provisioning logic of Flatcar OS / client services or by our users.
+   This includes validating signatures of our images by our users / their provisioning automation, validating update payloads by our update client, and verifying integrity during the boot process.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:65]:
```suggestion
   This includes validating signatures of Flatcar images by end users (or their provisioning automation), validating update payloads by the Flatcar update client, and verifying integrity during the boot process.
```

File: docs/reference/supply-chain.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+
+#### Foundation
+
+Flatcar builds its supply chain security on a number of basic concepts which, in summary, provide the foundation for securing our supply chain.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:70]:
```suggestion
Flatcar builds its supply chain security on a number of basic concepts which, in summary, provide the foundation for securing the entire Flatcar supply chain.
```

File: docs/reference/supply-chain.md [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+These foundational concepts fall into one of the two areas lined out above.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:71]:
```suggestion
These foundational concepts fall into one of the two areas outlined above.
```

File: docs/reference/supply-chain.md [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+Building on the foundational concepts outlined above Flatcar builds its supply chain security.
+In this section we will discuss the overall Flatcar build and release process as well as user-side provisioning, update, and operation - with a special focus on read models and supply chain security.
+
+##### OS Image build and release of a new OS version
+
+Flatcar builds are reproducible; the software configuration state of any given release (or even nightly build) is recorded in git repositories and can be reproduced by a simple git clone + checkout + rebuild.
+We employ a number of mechanisms to make this process tamper-proof and to make artifacts we produce attestable.
+
+Flatcar release images and related artifacts are signed at build time (on the secure build infrastructure) with a 4096 bit GPG RSA key.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:99]:
```suggestion
Flatcar release images and related artifacts are automatically signed at build time (on the secure build infrastructure) with a 4096 bit GPG RSA key.
```

Comment (ahrkrak) [reply docs/reference/supply-chain.md:99]: (if this is correct?)

File: docs/reference/supply-chain.md [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+Access to the image signing key is restricted to core maintainers.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:100]: where is the concept of "core maintainer" defined? does it map to a github team?

File: docs/reference/supply-chain.md [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+The image signing key is always stored encrypted and has a lifetime of one year.
+Renewing the image signing key requires split secrets of multiple maintainers.
+
+![supply_chain_build](../img/supply-chain-build.png)
+
+###### Inputs
+1. Flatcar's build automation and package definition repositories.
+   Write access to repositories is limited to trusted group of core Flatcar maintainers.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:108]: where is this set of "trusted" group or core maintainers defined? is that a subset of the core maintainers mentioned above? (are the other core maintainers not trusted?)

File: docs/reference/supply-chain.md [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+   All changes are reviewed by at least one maintainer before merge.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:109]: is this enforced? or more of a guideline?

File: docs/reference/supply-chain.md [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+      Package definitions include the [coreos-overlay](https://github.com/flatcar-linux/coreos-overlay) and [portage-stable](https://github.com/flatcar-linux/portage-stable) repositories.
+2. Upstream source tarballs of applications and libraries shipped with Flatcar.
+   Secured by cryptographic checksums stored in Flatcar's build automation repos (Gentoo standard).
+3. The SDK container.
+   The container is the result of a previous build and is validated by its container registry checksum.
+
+###### Process
+
+The OS image (and optionally, SDK) build and artifact signing is performed on hardened, secure build infrastructure.
+Access to the infrastructure is limited to a small number of core maintainers and is reviewed regularly.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:122]: is this a subset of the "core maintainers" mentioned above? if so, what is that subset? how's it defined?

Comment (ahrkrak) [reply docs/reference/supply-chain.md:122]: (in other words, to be more explicit, I think these separate sets of people need defining somewhere)

File: docs/reference/supply-chain.md [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+   2. Package definition (ebuild) repositories "pinned" as git submodules to a commit in the build automation repo.
+      Package definitions include the [coreos-overlay](https://github.com/flatcar-linux/coreos-overlay) and [portage-stable](https://github.com/flatcar-linux/portage-stable) repositories.
+2. Upstream source tarballs of applications and libraries shipped with Flatcar.
+   Secured by cryptographic checksums stored in Flatcar's build automation repos (Gentoo standard).
+3. The SDK container.
+   The container is the result of a previous build and is validated by its container registry checksum.
+
+###### Process
+
+The OS image (and optionally, SDK) build and artifact signing is performed on hardened, secure build infrastructure.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:121]: what does "hardened, secure" actually mean? I read that as fluff without more detail. a dedicated machine (not a VM) in a secure, access-controlled Equinix Metal data center -- now that's more specific.

File: docs/reference/supply-chain.md [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+   Integrity of source tarballs is validated against multiple cryptographic checksums stored in package definition (ebuild) repos.
+3. Building of apps and libraries, and generation of installation images and update image.
+   Optionally, a new SDK is built prior to the OS apps and libs.
+   This is done only for new major Alpha releases.
+4. Signing of artifacts to enable validation of authenticity at provisioning time.
+   1. A verity hash of the OS partition is generated and injected into the initrd so Flatcar can verify tamper-free OS partition at boot time.
+   2. There is an extra layer of security for the update image.
+      Many Flatcar deployments use automated updates so special care is take to ensure these are not compromised.
+      A core maintainer downloads the update image from the secure build server, and validates the image and its server signature.
+      The image is then signed with a key stored on a HSM, in an air-gapped environment so the key is never exposed.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:137]:
```suggestion
      The image is then signed with a key stored on a hardware security module (HSM), in an air-gapped environment so the key is never exposed to the internet. The HSM is stored in a combination-locked safe.
```

File: docs/reference/supply-chain.md [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+Accounts with access to the update server use 2-factor authentication.
+
+1. Artifacts and signatures are uploaded from the secure build infrastructure to the public image server.
+   The SDK container (if applicable) is pushed to the container registry (Flatcar uses GHCR at the time of writing).
+2. The update image and its (manually generated) signature are uploaded to the update server by the person who performed the manual signing step.
+
+##### Provisioning-time / OS upgrade / run-time
+
+Flatcar ships with a number of mechanisms to attest the authenticity of artifacts consumed by our users when provisioning and when updating Flatcar.
+We further assess the authenticity of all OS binaries - which reside on a separate, read-only partition - at each boot.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:152]:
```suggestion
Further, Flatcar assesses the authenticity of all OS binaries - which reside on a separate, read-only partition - at each boot.
```

Comment (ahrkrak) [reply docs/reference/supply-chain.md:152]: (is this the case? can you say anything more specific than "flatcar" here? which part of flatcar does that authenticity check? how?)

File: docs/reference/supply-chain.md [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+The OS partition is validated on each boot using `dm-verity`. The verity hash is baked into the init-rd at build time.
+
+
+### Future improvements
+
+To further enhance attestability and supply chain security we consider the below (non-exhaustive) list of improvements for Flatcar in the future.
+
+#### SLSA
+
+- Provide user-verifiable provenance of Flatcar's supply chain security.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:194]: Needs more text about how we're aiming to meet SLSA Level 4, and all the steps required to do that, and that we're actively working on that.

File: docs/reference/supply-chain.md [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+###### Validation of OS partition at boot time
+
+All operating system binaries are contained in a separate, immutable (read-only) partition which is mounted to `/usr` at system boot.
+No OS binaries exist outside `/usr` and no individual files can be changed.
+The OS partition is validated on each boot using `dm-verity`. The verity hash is baked into the init-rd at build time.
+
+
+### Future improvements
+
+To further enhance attestability and supply chain security we consider the below (non-exhaustive) list of improvements for Flatcar in the future.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:190]: create github issues for each and link from here

File: docs/reference/supply-chain.md (hunk 7 above)
Comment (t-lo) [reply docs/reference/supply-chain.md:100]: For Flatcar, yes. Core maintainers / trusted individuals are mentioned in SLSA without a clear definition (as the standard is work-in-progress).

File: docs/reference/supply-chain.md (hunk 11 above)
Comment (t-lo) [reply docs/reference/supply-chain.md:121]: Good point, will rephrase.

File: docs/reference/supply-chain.md (hunk 12 above)
Comment (t-lo) [reply docs/reference/supply-chain.md:137]: The combination lock does not apply, everything else makes sense. Will rephrase.

File: docs/reference/supply-chain.md [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+
+###### Outputs
+
+After image builds conclude, OS images, update image, related artifacts and signature files reside on the secure build infrastructure and are ready for publishing.
+Access to the public image and update servers is limited to a small number of engineers.
+Accounts with access to the update server use 2-factor authentication.
+
+1. Artifacts and signatures are uploaded from the secure build infrastructure to the public image server.
+   The SDK container (if applicable) is pushed to the container registry (Flatcar uses GHCR at the time of writing).
+2. The update image and its (manually generated) signature are uploaded to the update server by the person who performed the manual signing step.
```
[the remark this answers is not in the document]
Comment (t-lo) [reply docs/reference/supply-chain.md:147]: We don't. This is the build infrastructure access lined out above. However, anyone without access to the HSM could not generate a valid signature, so the update would be rejected by clients.

File: docs/reference/supply-chain.md (hunk 9 above)
Comment (t-lo) [reply docs/reference/supply-chain.md:109]: It's a process we follow but do not enforce (though SLSA does not state anything about enforcing). We have a tracking issue to enforce this, which in turn will need some modifications to our build automation: https://github.com/flatcar-linux/Flatcar/issues/815

**Comment by t-lo** [comment]
@ahrkrak Ready for another round. I elaborated on the SLSA l4 gaps and ensured we have tracking tickets for all these.

File: docs/reference/supply-chain.md [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+3. builds from a modified source
+4. a compromised build process
+5. use of a compromised dependency
+6. publishing of a compromised package or image
+7. a compromised package or image repository
+8. injection / use of a compromised package or image
+
+To counter these threats, SLSA defines [requirements](https://slsa.dev/spec/v0.1/requirements) for sources, builds, and provenance, as well as common (overall) requirements.
+This section will summarise the requirements and relate the SLSA levels of Flatcar's implementation.
+
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:30]: I don't think O unambiguously means "doesn't support" - I think "–" more clearly suggests that. And added note numbers in the table.

File: docs/reference/supply-chain.md [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+1. Installation images contain a full, pre-partitioned disk including the full OS.
+   No additional OS binaries or packages are installed during or after provisioning.
+   (The root partition, which does not contain OS artifacts, is resized to span the whole disk at provisioning time for user convenience.)
+2. All OS binaries and libraries reside on a separate _read-only_ partition (mounted to `/usr` at runtime).
+   The partition cannot be written to.
+
+
+#### Flatcar supply chain security mechanisms
+
+Building on the foundational concepts outlined above Flatcar builds its supply chain security.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:103]:
```suggestion
Flatcar builds its supply chain security on the foundational concepts outlined above.
```

File: docs/reference/supply-chain.md (hunk 9 above)
Comment (ahrkrak) [reply docs/reference/supply-chain.md:122]: OK, can leave this text as is then. Would be good to enforce it if possible (and I'd personally like SLSA to specify enforcement).

File: docs/reference/supply-chain.md (hunk 17 above)
Comment (ahrkrak) [reply docs/reference/supply-chain.md:30]: Also, shouldn't the table have a key? Even though it's relatively simple - see the Suse doc for an example.

File: docs/reference/supply-chain.md [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+
+###### Validation of update images at OS upgrade time
+
+Flatcar ships with a mechanism to auto-upgrade itself to new releases.
+The client service, `update_engine`, is included in the OS partition of the Flatcar image, i.e. it is on a read-only, `dm-verity`-validated partition.
+Before installation, update images are validated against the update signing key - this key has a separate, even stronger security process than the image signing key (see build process for details).
+`update_engine` uses a baked-in public key for validation.
+An update is installed only after successful validation.
+
+### Future improvements
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:208]: Where is the plan to address the "Common - Superusers" gap?

File: docs/reference/supply-chain.md [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+| Common - Superusers                               |              |              |              |      ✓       |       ○       |
+
+
+**Notes**
+
+1. Build integrity - Hermetic builds: While Flatcar includes the potential for hermetic builds today - all sources are known in advance and can be staged to a build machine isolated from the network - the current build infrastructure and automation does not implement this feature.
+   A [tracking issue](https://github.com/flatcar-linux/Flatcar/issues/833) exists to address this in the future.
+2. Build integrity - Reproducible: Many software packages such as compilers and core libraries insert build-variable information such as timestamps, user IDs, and host names into their binaries during the build process.
+   While Flatcar's builds are 100% reproducible, the output may differ in a bit-by-bit comparison (as defined by SLSA) ONLY in places where this volatile information is compiled into the binaries.
+3. Common - Security: The SLSA requirement is TBD, hence not well defined.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:61]: Are we being too harsh on ourselves here? Suse claim compliance with the following statement: "Red team scans for vulnerabilities, build team patches build environment. Machine identity is enforced, user isolation by policy."

I'm not sure TPM is in scope, although we might want to flag it as an area we're working on to go BEYOND the minimum requirements of SLSA...

**Review by ahrkrak — CHANGES_REQUESTED** [review]

File: docs/reference/supply-chain.md (hunk 17 above)
Comment (t-lo) [reply docs/reference/supply-chain.md:30]: I used "o" for "best effort" - builds are reproducible but volatile parts of the binary (like timestamps and users) change, and we don't currently have tooling in place that honors this (and skips respective sections of the binaries). I've merged your proposed "-" for the time being.

By key you mean anchor, so we can link it?

File: docs/reference/supply-chain.md (hunk 20 above)
Comment (t-lo) [reply docs/reference/supply-chain.md:61]: Though we ensure common security practices we don't proactively validate security of our infrastructure (e.g. by making efforts to compromise it). 
And, while unfinished, the SLSA requirements strongly hint at a verifiable secure boot chain without gaps (i.e. extending attestability to the OS hosting the builds). SuSE's statement does not include any of this and seems very weak; it's the first thing I'd attack when reading their self-declaration critically.

File: docs/reference/supply-chain.md (hunk 19 above)
Comment (t-lo) [reply docs/reference/supply-chain.md:208]: That's a larger endeavor that needs its own design phase - almost like a roadmap item - so a tracking issue won't cut it. but it's clearly missing here. I will add a paragraph on this.

File: docs/reference/supply-chain.md [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+2. Generate additional provenance for the whole image: [tracking issue](https://github.com/flatcar-linux/Flatcar/issues/814)
+
+#### Build time
+
+1. Make release builds hermetic by providing all required assets beforehand and isolating the build machine from the network during build, to address the "Build integrity - Hermetic" requirement
+   ([tracking issue](https://github.com/flatcar-linux/Flatcar/issues/833)).
+2. Establish a secure boot chain using TPM support when it becomes available (see "Provisioning-time" item 2. below).
+3. Remove login (local and remote) from build infrastructure and automate all build infra properties (infra-as-code).
+   Require approval from 2 administrators for every change.
+   This will address the SLSA Security / Superusers requirement.
```
Comment (ahrkrak) [inline docs/reference/supply-chain.md:227]:
```suggestion
   This will address the "Security - Superusers" requirement.
```

**Review by ahrkrak — APPROVED** [review]
lgtm

**Pull request merged**